File: 0057-insert-interval/0057-insert-interval.cpp
class Solution {
private:
    vector<vector<int>> mergeIntervals(vector<vector<int>>& intervals)
    {
        vector<vector<int>> res;
        
        auto& currInterval=intervals[0];
        
        for(auto& interval : intervals)
        {
            if(currInterval[1] >= interval[0])
            {
               currInterval[1]=max(currInterval[1],interval[1]); 
            }
            else
            {
                res.push_back(currInterval);
                currInterval=interval;
            }
        }
        
        res.push_back(currInterval);
        
        return res;
    }
    
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        
        if(intervals.size() == 0)
            return {newInterval};
             
        if(newInterval.size() == 0)
            return intervals;
        
        vector<vector<int>> res;
        
        for(auto& interval : intervals)
        {
            if(newInterval.size() != 0 && interval[0] > newInterval[0])
            {
                res.push_back(newInterval);
                res.push_back(interval);
                newInterval={};
            }
            else
            {
                res.push_back(interval); 
            }            
        }
        
        if(newInterval.size() != 0)
            res.push_back(newInterval);
        
        return mergeIntervals(res);
    }
};

File: 0057-insert-interval/0057-insert-interval.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        int n = intervals.length;
        
        ArrayList<int[]> res = new ArrayList<>();
        
        int index = 0;
        
        while(index < n && intervals[index][1] < newInterval[0]) {
            res.add(intervals[index]);
            index++;
        }
        
        while(index < n && intervals[index][0] <= newInterval[1]) {
            newInterval[0] = Math.min(newInterval[0], intervals[index][0]);
            newInterval[1] = Math.max(newInterval[1], intervals[index][1]);
            index++;
        }
        
        res.add(newInterval);
        
        while(index < n) {
            res.add(intervals[index]);
            index++;
        }
        
        int[][] mergedIntervals = new int[res.size()][2];
        
        for(int i = 0; i < res.size(); i++) {
            mergedIntervals[i] = res.get(i);
        }
        
        return mergedIntervals;
    }
}